com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -0,0 +1,217 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using MLAPI.Configuration;
+using MLAPI.Messaging;
+using MLAPI.NetworkVariable;
+using MLAPI.Serialization;
+using MLAPI.Serialization.Pooled;
+using MLAPI.Transports;
+using UnityEngine;
+using UnityEngine.UIElements;
+
+namespace MLAPI
+{
+    internal struct Entry
+    {
+        public ulong m_NetworkObjectId; // the NetworkObjectId of the owning GameObject
+        public ushort m_Index; // the index of the variable in this GameObject
+        public ushort m_Position; // the offset in our m_Buffer
+        public ushort m_Length; // the length of the data in m_Buffer
+
+        public const int k_NotFound = -1;
+    }
+
+    internal class EntryBlock
+    {
+        private const int k_MaxVariables = 64;
+
+        public Entry[] m_Entries = new Entry[k_MaxVariables];
+        public int m_LastEntry = 0;
+
+        public int Find(ulong networkObjectId, int index)
+        {
+            for (int i = 0; i < m_LastEntry; i++)
+            {
+                if (m_Entries[i].m_NetworkObjectId == networkObjectId && m_Entries[i].m_Index == index)
+                {
+                    return i;
+                }
+            }
+
+            return Entry.k_NotFound;
+        }
+
+        public int AddEntry(ulong networkObjectId, int index)
+        {
+            var pos = m_LastEntry++;
+            var entry = m_Entries[pos];
+
+            entry.m_NetworkObjectId = networkObjectId;
+            entry.m_Index = (ushort)index;
+            entry.m_Position = 0;
+            entry.m_Length = 0;
+            m_Entries[pos] = entry;
+
+            return pos;
+        }
+
+    }
+
+    public class SnapshotSystem : INetworkUpdateSystem, IDisposable
+    {
+        private EntryBlock m_Snapshot = new EntryBlock();
+        private EntryBlock m_ReceivedSnapshot = new EntryBlock();
+
+        byte[] m_Buffer = new byte[20000];
```

Comment:
Shouldn't this buffer be part of the snapshot? Basically there should be a snapshot struct with an entryblock and a buffer.

Reply:
Yes. The buffer area will get segregated into two areas. I'll definitely comment on that.